### DISCUSSION TITLE
[Berry] Replace single line in file

### DISCUSSION THREAD

**Discussion by Slyke** [arendst/Tasmota#15874] [Q&A] [answered]
This is related to https://github.com/arendst/Tasmota/discussions/15806

I have a file structure like this:
```
[
  {"hasBeenUploaded": true, "someData": "aboutSomeStuff1"},
  {"hasBeenUploaded": true, "someData": "aboutSomeStuff2"},
  {"hasBeenUploaded": true, "someData": "aboutSomeStuff3"},
  {"hasBeenUploaded": false, "someData": "aboutSomeStuff4"},
  {"hasBeenUploaded": false, "someData": "aboutSomeStuff5"},
  {"hasBeenUploaded": false, "someData": "aboutSomeStuff6"},
  {"hasBeenUploaded": false, "someData": "aboutSomeStuff7"}
]
```

I'm iterating through each line, parsing it into JSON, and then uploading the JSON object to a testing server.

I'm trying to figure out a way to either remove the line that was just uploaded, or change the `hasBeenUploaded` property to `true` and reinsert into the file, depending on an option being set.

I've tried replacing the line with an empty string or spaces, but it either inserts the spaces or overwrites what's currently there (so nothing is shifted backwards).

Is there an easy way to do this without reading in the entire file, or shifting all future lines back by 1?

I store the start and end of reach line into variables:
```
def every_second()
    recordUploadFile = open('/sd/log.json', 'r+')

    recordUploadFile.seek(self.uploadCursorPos)
    var lineStartCursorPos = self.uploadCursorPos
    var lineBuf = recordUploadFile.readline()
    self.uploadCursorPos = recordUploadFile.tell()
    var lineEndCursorPos = self.uploadCursorPos
    recordUploadFile.close()
```

**Comment by sfromis** [comment, marked as answer]
If you want bytes moved around, that inherently at least means reading/writing the rest of the file. Magic does not happen.

Sure, using seek and write, you can replace the exact same number of bytes at some position, meaning that you can create the initial data to have the same exact length as the intended updates.

I'd be inclined to a different more general strategy, avoiding "too big" files, by instead splitting into files small enough to reasonably be able to load the JSON into a map, do whatever update there, and dump the map replacing that one file. While this strategy is obviously a bit less "efficient", it is much more resilient.

**Reply by Slyke** [reply]
Your suggestion gave me an idea; I think I'll create a temporary file that has the "new" file data and just insert each line until EoF on the original, then delete the original and rename the temporary one. It would potentially cause data to be reuploaded if upload is interrupted before finishing, but should be okay as it'll have a RTC.

I'm making a sensor logger that may or may not have connectivity. If it's configured to do so, it should upload all unuploaded data if it detects and connects to a specific wifi network.

**Reply by s-hadinger** [reply]
Doing so many writes/rewrites to the internal flash is dangerous and may lead drastically reducing the life of flash memory. There might be some libraries out there optimized to ensure wear leveling.

**Reply by Slyke** [reply]
Yep! That is true! I'm actually using an SD Card for logging and temp file, so not too worried.

**Reply by sfromis** [reply]
SD cards are not great either, it is a known issue that Pi users are wearing out their SD cards by treating it as if it was a SSD.
